src/format/mod.rs

```diff
@@ -0,0 +1,888 @@
+//! Module containing the formatting logic.
+
+mod assert;
+mod utils;
+mod week;
+mod write;
+
+use core::fmt;
+use core::num::IntErrorKind;
+use core::str;
+
+use bitflags::bitflags;
+
+use crate::{Error, Time};
+use assert::{assert_sorted, assert_sorted_elem_0, assert_to_ascii_uppercase};
+use utils::{Cursor, SizeLimiter};
+use week::{iso_8601_year_and_week_number, week_number, WeekStart};
+use write::Write;
+
+/// Alias to a `c_int`.
+#[cfg(feature = "std")]
+type Int = std::os::raw::c_int;
+/// Fallback alias to a `c_int`.
+#[cfg(not(feature = "std"))]
+type Int = i32;
+
+/// List of weekday names.
+const DAYS: [&str; 7] = [
+    "Sunday",
+    "Monday",
+    "Tuesday",
+    "Wednesday",
+    "Thursday",
+    "Friday",
+    "Saturday",
+];
+
+/// List of uppercase weekday names.
+const DAYS_UPPER: [&str; 7] = [
+    "SUNDAY",
+    "MONDAY",
+    "TUESDAY",
+    "WEDNESDAY",
+    "THURSDAY",
+    "FRIDAY",
+    "SATURDAY",
+];
+
+/// List of month names.
+const MONTHS: [&str; 12] = [
+    "January",
+    "February",
+    "March",
+    "April",
+    "May",
+    "June",
+    "July",
+    "August",
+    "September",
+    "October",
+    "November",
+    "December",
+];
+
+/// List of uppercase month names.
+const MONTHS_UPPER: [&str; 12] = [
+    "JANUARY",
+    "FEBRUARY",
+    "MARCH",
+    "APRIL",
+    "MAY",
+    "JUNE",
+    "JULY",
+    "AUGUST",
+    "SEPTEMBER",
+    "OCTOBER",
+    "NOVEMBER",
+    "DECEMBER",
+];
+
+// Check day and month tables
+const _: () = {
+    assert_to_ascii_uppercase(&DAYS, &DAYS_UPPER);
+    assert_to_ascii_uppercase(&MONTHS, &MONTHS_UPPER);
+};
+
+bitflags! {
+    /// Formatting flags.
+    struct Flags: u32 {
+        /// Use left padding, removing all other padding options in most cases.
+        const LEFT_PADDING = 1 << 0;
+        /// Change case for a string value.
+        const CHANGE_CASE  = 1 << 1;
+        /// Convert a string value to uppercase.
+        const UPPER_CASE   = 1 << 2;
+    }
+}
+
+impl Flags {
+    /// Check if one of the case flags is set.
+    fn has_change_or_upper_case(self) -> bool {
+        let flag = Flags::CHANGE_CASE | Flags::UPPER_CASE;
+        !self.intersection(flag).is_empty()
+    }
+}
+
+/// Padding method.
+#[derive(Debug, Default, Copy, Clone, Eq, PartialEq)]
+enum Padding {
+    /// Left padding.
+    #[default]
```

Comment:
I'm confused how this works on the MSRV 1.56.0 build since I think `#[default]` on enums was only stabilized in 1.62.0.

Reply:
Found the bug in the build. It will be fixed here:

- https://github.com/artichoke/strftime-ruby/pull/13

It looks like we should either remove the `Default` impl or impl it manually.

Reply:
I would want to have at least Rust 1.58, since the code uses heavily the [format arguments capture](https://blog.rust-lang.org/2022/01/13/Rust-1.58.0.html#captured-identifiers-in-format-strings) feature.

Reply:
sounds good. I can fix this up after the merge if need be.

Reply:
Filed:

- https://github.com/artichoke/strftime-ruby/pull/14

Fix is here:

- https://github.com/artichoke/strftime-ruby/pull/15